File: frameworks/dotnet/fastendpoints/UnitTests/ClientTests.cs
using Client.Kiota;
using Client.Kiota.Models;
using Microsoft.Kiota.Abstractions.Authentication;
using Microsoft.Kiota.Abstractions.Serialization;
using Microsoft.Kiota.Http.HttpClientLibrary;

namespace UnitTests;

/// <summary>
/// The Kiota client in Client/, which FastEndpoints.OpenApi.Kiota generates from the document
/// FastEndpoints.OpenApi writes, calling the Implementation in the test host. These hold the
/// client to what FastEndpoints answers, not FastEndpoints to a corpus row, so they carry no
/// corpus trait.
/// </summary>
public sealed class ClientTests : TestBase<App>
{
    private readonly FastEndpointsClient client;

    public ClientTests(App app)
    {
        HttpClient http = app.CreateClient();
        // Building the client registers Kiota's JSON factories, which KiotaJsonSerializer reads.
        client = new FastEndpointsClient(new HttpClientRequestAdapter(new AnonymousAuthenticationProvider(), httpClient: http)
        {
            BaseUrl = http.BaseAddress!.ToString().TrimEnd('/'),
        });
    }

    [Fact]
    public async Task Json_small_is_the_payload()
    {
        Payload? answer = await client.Json.Small.GetAsync(cancellationToken: Cancellation);

        JsonNode actual = JsonNode.Parse(await KiotaJsonSerializer.SerializeAsStringAsync(answer!, serializeOnlyChangedValues: false, cancellationToken: Cancellation))!;
        Assert.True(JsonNode.DeepEquals(Expected.Json("items.small.json"), actual), actual.ToJsonString());
    }

    [Fact]
    public async Task A_read_item_is_typed_because_the_endpoint_names_its_response()
    {
        Item? answer = await client.Items[17].GetAsync(cancellationToken: Cancellation);

        Assert.Equal(17, answer!.Id);
    }

    [Fact]
    public async Task A_replaced_item_is_typed()
    {
        ReplacedItem item = await ReadAsync("items.new.json", ReplacedItem.CreateFromDiscriminatorValue);

        Item? answer = await client.Items[17].PutAsync(item, cancellationToken: Cancellation);

        Assert.Equal(17, answer!.Id);
        Assert.Equal(item.Name, answer.Name);
    }

    [Fact]
    public async Task A_validated_order_is_bound()
    {
        CheckedOrder order = await ReadAsync("order.small.json", CheckedOrder.CreateFromDiscriminatorValue);

        Bound? answer = await client.Body.Validate.Small.PostAsync(order, cancellationToken: Cancellation);

        Assert.Equal(4, answer!.Fields);
        Assert.Equal(order.CustomerId, answer.Echo!.CustomerId);
    }

    [Fact]
    public async Task A_rejected_order_is_FastEndpoints_ErrorResponse_thrown_as_an_exception()
    {
        CheckedOrder order = await ReadAsync("order.invalid.json", CheckedOrder.CreateFromDiscriminatorValue);

        ErrorResponse refused = await Assert.ThrowsAsync<ErrorResponse>(() => client.Body.Validate.Small.PostAsync(order, cancellationToken: Cancellation));

        Assert.Equal(400, refused.ResponseStatusCode);
        Assert.Equal(["customerId", "status", "lines"], refused.Errors!.AdditionalData.Keys);
    }

    [Fact]
    public async Task Query_and_path_parameters_are_typed()
    {
        EchoedOfQueryOne? query = await client.Query.One.GetAsync(c => c.QueryParameters.Page = 417, Cancellation);
        EchoedOfParametersTwo? path = await client.Parameters[4821].WithSecond[7390].GetAsync(cancellationToken: Cancellation);

        Assert.Equal("small", query!.Size);
        Assert.Equal("small", path!.Size);
    }

    /// <summary>A committed payload, read into the client's own model.</summary>
    private static async Task<T> ReadAsync<T>(string file, ParsableFactory<T> factory) where T : IParsable =>
        (await KiotaJsonSerializer.DeserializeAsync(File.ReadAllText(Path.Combine(Expected.Directory, file)), factory, TestContext.Current.CancellationToken))!;
}

File: frameworks/java/javalin/UnitTests/src/test/java/unittests/ClientTests.java
package unittests;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.List;
import java.util.Map;

import client.ApiClient;
import client.ApiException;
import client.api.DefaultApi;
import client.model.Bound;
import client.model.Echoed;
import client.model.Item;
import client.model.Line;
import client.model.OrderRequest;
import client.model.Payload;
import io.javalin.testtools.HttpClient;
import io.javalin.testtools.JavalinTest;
import org.junit.jupiter.api.Test;

/**
 * The client in Client/, which OpenAPI Generator writes from the document the @OpenApi annotations
 * describe, calling the Implementation JavalinTest started. These hold the client to what Javalin
 * answers, not Javalin to a corpus row, so they carry no corpus tag.
 */
class ClientTests extends JavalinApp {

    @Test
    void jsonSmallIsTheFirstRow() {
        JavalinTest.test(app(), (server, client) -> {
            Payload answer = api(client).jsonSmallGet();

            assertEquals("small", answer.getSize());
            assertEquals(1, answer.getCount());
            assertEquals(1, answer.getItems().getFirst().getId());
        });
    }

    @Test
    void aRowIsTypedByTheResponseTheAnnotationNames() {
        JavalinTest.test(app(), (server, client) -> {
            Item row = api(client).itemsIdGet(17);

            assertEquals(17, row.getId());
        });
    }

    @Test
    void aValidatedOrderIsBound() {
        JavalinTest.test(app(), (server, client) -> {
            Bound answer = api(client).bodyValidateSmallPost(order(1, "open", 1, 1));

            assertEquals(4, answer.getFields());
            assertEquals(1, answer.getEcho().getCustomerId());
        });
    }

    @Test
    void aRejectedOrderIsAnApiExceptionWithTheStatus() {
        JavalinTest.test(app(), (server, client) -> {
            ApiException refused = assertThrows(ApiException.class, () -> api(client).bodyValidateSmallPost(order(0, "", 0, 0)));

            assertEquals(400, refused.getCode());
        });
    }

    /** The echo is a type parameter of Echoed, which the document names as an object, so the client reads it as a map. */
    @Test
    void queryAndPathParametersAreTyped() {
        JavalinTest.test(app(), (server, client) -> {
            Echoed query = api(client).queryOneGet(417);
            Echoed path = api(client).parametersOneWithSecondTwoGet(4821, 7390);

            assertEquals(Map.of("page", 417.0), query.getEcho());
            assertEquals(Map.of("one", 4821.0, "two", 7390.0), path.getEcho());
        });
    }

    private static DefaultApi api(HttpClient client) {
        ApiClient api = new ApiClient();
        api.setBasePath(client.getOrigin());
        return new DefaultApi(api);
    }

    /** An order of one line, in the client's own model. */
    private static OrderRequest order(int customerId, String status, int productId, int qty) {
        return new OrderRequest().customerId(customerId).status(status).lines(List.of(new Line().productId(productId).qty(qty)));
    }
}
